src/pcre2_jit_compile.c

```diff
@@ -5888,29 +5969,32 @@ while (TRUE)
     case OP_NOTEXACT:
     case OP_NOTEXACTI:
 #if defined SUPPORT_UNICODE && PCRE2_CODE_UNIT_WIDTH != 32
-    if (common->utf) return consumed;
+    if (common->utf)
+      {
+      chars_end = chars;
+      continue;
+      }
 #endif
     any = TRUE;
     repeat = GET2(cc, 1);
     cc += 1 + IMM2_SIZE + 1;
     break;
 
     default:
-    return consumed;
+    chars_end = chars;
+    continue;
     }
 
+  SLJIT_ASSERT(chars < chars_end);
+
   if (any)
```

Comment:
Here's a suggestion: as far as I can see, two out of three of the uses of recursion in the original code were completely unnecessary.

There was OP_QUERY, OP_CRSTAR/OP_CRQUERY, and OP_BRA which do recursion. The first two don't need it!

For example, OP_QUERY could be:

```
switch (*cc) {

case OP_QUERY:
add_prefix_chars(*cc, caseless, repeat=1);
cc += len;
continue;
```

Something like that. Basically, take the blocks of code underneath the switch (`if (any) ...` and `if (class) ...` and the code for handling a single character). Lift those out into functions. Then you can process the OP_QUERY immediately, and just carry on to the next opcode, no need for any funky recursion tricks.